File: src/Julia/prefix-julia-plus/decode.jl
function decode(c :: Word , q :: Key) :: Word
    k = deepcopy(q)
    p :: Word = []
    m = 0
    i = 1
    while i <= length(c)
        for r in k
            if r.mode == m && prefix(r.writes,c[i:end])
                #push!(c,r.writes)
                p = vcat(p, r.reads)
                i += length(r.writes)
                m = r.goes
                @goto next_mode
            end
        end
        @label next_mode
    end
    p
end

File: src/Go/pre-go/decode.go
package main
import "strings"
func (m mode) decode(x string) (X, y string, g int) {
	for _ , h := range m {
		if strings.HasPrefix(x,h.writes) {
			X = x[len(h.writes):]
			y = h.reads
			g = h.goes
			return
		}
	}
	return
}
func (f filter) decode(x string) string {
	current_mode := 0
	y := ""
	z := ""
	for ;x != ""; {
		x,z, current_mode = f[current_mode].decode(x)
		y = y + z
	} 
	return y
}